Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -647,8 +647,18 @@ extension SourceKitServer {
 
   func shutdown(_ request: Request<ShutdownRequest>) {
     _prepareForExit()
+    let sema = DispatchSemaphore(value: 0)
     for service in languageService.values {
-      service.shutdown()
+      service.shutdown() {
+        sema.signal()
+      }
+    }
+    // Wait for all services to shut down before sending the shutdown response.
+    // Otherwise we might terminate sourcekit-lsp while it still has open
+    // connections to the toolchain servers, which could send messages to
+    // sourcekit-lsp while it is being deallocated, causing crashes.
+    for _ in languageService.values {
+      sema.wait()
```

Comment:
Better to use a DispatchGroup instead of multiple waiting on a counting semaphore.  Also, is there a reason you want to block the server queue, or would it be okay to have the rest of this code run async? You can have the DispatchGroup call you back instead of blocking.

Reply:
DispatchGroup is definitely the better solution. I didn’t know about it.